src/api-extension/cart.service.ts

```diff
@@ -203,28 +203,71 @@ export class CartService {
 
     coupons = this.filterCouponsByLimit(coupons, couponsLimit);
 
-    const validatedCoupons =
+    let validatedCoupons =
       await this.voucherifyConnectorService.validateStackableVouchersWithCTCart(
         coupons.filter((coupon) => coupon.status != 'DELETED'),
         cart,
         this.productMapper.mapLineItems(cart.lineItems),
         sessionKey,
       );
 
+    const productsToAdd = await convertUnitTypeCouponsToFreeProducts(
+      validatedCoupons,
+      this.commerceToolsConnectorService.getClient(),
+      this.getPriceSelectorFromCart(cart),
+    );
+
+    const productsToChange = productsToAdd.filter(
+      (product) => product.discount_difference !== 0,
+    );
```

Comment:
discount_difference could be probably simplified to Boolean without difference calculation. Instead You could check is there is any `currentPriceAmount` and if `currentPriceAmount` is different than `freeItem?.applied_discount_amount`

Reply:
fixed